services/slam/builtin/builtin_test.go

```diff
@@ -966,46 +967,100 @@ func TestSLAMProcessSuccess(t *testing.T) {
 
 	createFakeSLAMLibraries()
 
-	attrCfg := &builtin.AttrConfig{
-		Sensors:          []string{"good_color_camera"},
-		ConfigParams:     map[string]string{"mode": "mono", "test_param": "viam"},
-		DataDirectory:    name,
-		MapRateSec:       &validMapRate,
-		DataRateMs:       validDataRateMS,
-		InputFilePattern: "10:200:1",
-		Port:             "localhost:4445",
-	}
-
-	// Create slam service
 	logger := golog.NewTestLogger(t)
-	grpcServer := setupTestGRPCServer(attrCfg.Port)
-	svc, err := createSLAMService(t, attrCfg, "fake_orbslamv3", logger, false, true)
-	test.That(t, err, test.ShouldBeNil)
 
-	slamSvc := svc.(internal.Service)
-	processCfg := slamSvc.GetSLAMProcessConfig()
-	cmd := append([]string{processCfg.Name}, processCfg.Args...)
-
-	cmdResult := [][]string{
-		{slam.SLAMLibraries["fake_orbslamv3"].BinaryLocation},
-		{"-sensors=good_color_camera"},
-		{"-config_param={mode=mono,test_param=viam}", "-config_param={test_param=viam,mode=mono}"},
-		{"-data_rate_ms=200"},
-		{"-map_rate_sec=200"},
-		{"-data_dir=" + name},
-		{"-input_file_pattern=10:200:1"},
-		{"-port=localhost:4445"},
-		{"--aix-auto-update"},
-	}
+	t.Run("Test online SLAM process", func(t *testing.T) {
+		deleteProcessedData := false
```

Comment:
Added a few additional test cases for checking the arg output of slam. These additional tests should cover the 6 potential values of delete_processed_data in both online and offline mode as well as check the default values/other values for various other arguments.

Reply:
Kat also brought up the option of splitting this check into a separate function so we could, in affect send in the config and it would return the deleteProcessedData variable. A similar process can be done for other attributes, however, it would require exposing additional functions such as `ExtractDeleteProcessedDataFromConfig` since testing is done in `builtin_test` and not the `builtin` package. 

If we went down this route, we could have these functions be only part of the internal Service, not the slam Service, similar to how we have `GetSLAMProcessBufferedLogReader`. However, I am hesitant to expose minor functions when we aren't even doing that for `getAndSaveDataDense/Sparse`.

Would love to hear your thoughts on this @tessavitabile and if we should do this now, have a test refactoring down the line or avoid doing this for the foreseeable future.

@kkufieta if I've left anything out please add to this note.

Reply:
Thanks for adding the additional test cases. One minor comment for the future is that I suggest having descriptive names for test cases instead of numbers, so that future maintainers can preserve the purpose of the test. You can also make these more targeting by only checking the parameter output you're interested in, i.e. just checking delete_processed_data in this case. If you want to take the time to do that now, I think it would improve maintainability, but it's not essential.

I don't think we need to add getters for the different process config parameters. Ideally, we would be able to create a stateless function GetSLAMProcessConfig that takes in service config parameters and outputs a process config. However, we handle defaults for service config parameters in NewBuiltIn that get used by GetSLAMProcessConfig, so we can't make that a stateless function.

Reply:
I think there might be a misunderstanding. What I suggested was to extract the logic that determines the value of `deleteProcessedData` into a function:

```go
deleteProcessedData := DetermineDeleteProcessedData(svcConfig.DeleteProcessedData)
```

This would allow us to test the logic by simply testing this function - instead of doing it indirectly via feeding in a config through `CreateSLAMService` --> `builtin.NewBuiltIn` --> `GetSLAMProcessConfig`. 

I think that would be more modular, easier to read, less code, and more maintainable. I don't see any downsides to doing that. Are there any?

Reply:
The only downside is that it exposes the function so that any user who imports slam can use it and we try to limit this as much as possible.

For an example of this see:
- `[component/base/wheeled_base](https://github.com/viamrobotics/rdk/blob/8244a190a24ddc8dd3937c9d99d2fcf6ad3da762/components/base/wheeled/wheeled_base.go#L224)`

However, I'm personally not against it but we'd probably want to do that for many other variables which would be out of scope for this PR.

Reply:
The "internal" directory in which we placed the `internal` package has been added to Golang to solve the problem we're discussing: We want to write well-structured modular code - especially as libraries grow large - and be able to import functions into other "internal" packages without exposing them to users who import our "main" package. (in our case, the "main" package I'm talking about is the `slam` package).

Any packages in the "internal" directory are not accessible to users who import slam. We happened to call the only package we placed in the "internal" directory folder `internal` - but it's not limited to just that. We could have multiple packages in the "internal" directory - so there's a lot of room for modularizing code any way we wish.

Source: https://go.dev/doc/go1.4#internalpackages

We should start modularizing code where we can and where it makes sense by utilizing the "internal" directory. In this case, creating a function, as I suggested, makes sense. Except what I suggested is not entirely correct, we need a second argument for it:

```go
deleteProcessedData := DetermineDeleteProcessedData(svcConfig.DeleteProcessedData, offlineFlag)
```

I agree that it might make sense to do this for other variables and code already implemented and that doing so is out of this PR's scope.

Reply:
Got it. Just extracting the logic for delete_processed_data (since it's more complicated than most parameters) makes sense to me. As Kat pointed out to me offline, we can avoid putting this in the exported interface by only added it to the internal package interface: https://github.com/viamrobotics/rdk/blob/main/services/slam/internal/slam_test_helper.go